utils/tail/tail.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tail implements "tail -F" functionality following rotated logs
+package tail
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/exp/inotify"
+)
+
+type Tail struct {
+	reader     *bufio.Reader
+	sync.Mutex // protects reader
+	filename   string
+	file       *os.File
+	stop       chan bool
+	watcher    *inotify.Watcher
+}
+
+const retryOpenInterval = time.Second
+const maxOpenAttempts = 3
+
+// NewTail starts opens the given file and watches it for deletion/rotation
+func NewTail(filename string) (*Tail, error) {
+	t := &Tail{
+		filename: filename,
+	}
+	var err error
+	t.stop = make(chan bool)
+	t.watcher, err = inotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("inotify init failed on %s: %v", t.filename, err)
+	}
+	t.Lock()
```

Comment:
To block readers before the file is open
On May 17, 2016 10:25 PM, "Tim St. Clair" <notifications@github.com> wrote:

> In utils/tail/tail.go
> <https://github.com/google/cadvisor/pull/1264#discussion_r63640227>:
>
> > +
> > +const retryOpenInterval = time.Second
> > +const maxOpenAttempts = 3
> > +
> > +// NewTail starts opens the given file and watches it for deletion/rotation
> > +func NewTail(filename string) (*Tail, error) {
> > +	t := &Tail{
> > +		filename: filename,
> > +	}
> > +	var err error
> > +	t.stop = make(chan bool)
> > +	t.watcher, err = inotify.NewWatcher()
> > +	if err != nil {
> > +		return nil, fmt.Errorf("inotify init failed on %s: %v", t.filename, err)
> > +	}
> > +	t.Lock()
>
> Why are you locking here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/cadvisor/pull/1264/files/8d80aaca81434fbf0804c6923f451df1869cb9c3#r63640227>
>